Faster consistency checker

tinwelint opened the pull request.

SRaghuram committed:
Faster consistency checker

Basically by reducing random access using an effient cache, checking only
parts of records through multiple passes. Also parallelizes some store
checks, like indexes.

tinwelint committed:
Removes more remnants of diff check

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/NodeField.java:
```diff
@@ -229,6 +251,13 @@ public void checkReference( RelationshipRecord relationship, NodeRecord node,
+                {
+                    if ( relationship.getFirstNode() != relationship.getSecondNode() )
+                    {
+                        cacheAccess.putToCacheSingle( node.getId(), 1, 1 );
```
First `1` should be changed to corresponding slot constant

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheTask.java:
```diff
@@ -0,0 +1,115 @@
+            {
+                if ( node.inUse() )
+                {
+                    fields[2] = node.getNextRel();
```
slot indexes in this class should be changed to constants

lutovich commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/Workers.java:
```diff
@@ -72,6 +72,18 @@ public Throwable await() throws InterruptedException
         return error;
     }
 
+    public Throwable awaitStrict()
```
Method not used

tinwelint replied:
No, but it's a general purpose method and matches `awaitAndThrowOnError` which also has a `strict` counterpart. I'm not too worried about it being there.

lutovich commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/CacheTask.java:
```diff
@@ -0,0 +1,115 @@
+import org.neo4j.kernel.impl.store.record.NodeRecord;
+
+/**
+ * Action to be run as part of a {@link CacheProcessor}.
```
`CacheProcessor` is not there any more

lutovich commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/ConsistencyCheckerTask.java:
```diff
@@ -0,0 +1,48 @@
+        this.numberOfThreads = threads;
+    }
+
+    public void stopScanning()
```
Method not used

lutovich commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/CountsBuilderDecorator.java:
```diff
@@ -207,27 +211,25 @@ public void check( NodeRecord record,
                 if ( record.inUse() )
                 {
+                    CacheAccess.Client client = records.cacheAccess().client();
+                    client.putToCacheSingle( record.getId(), 0, 1 );
```
Maybe use constants instead of 0 and 1

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/statistics/AccessStatistics.java:
```diff
@@ -0,0 +1,179 @@
+            appendStat( buf, "ScatterIndex", scatterIndex );
+            if ( scatterIndex > 0.5 )
+            {
+                buf.append( format( "%n *** Property Store reorgization is recommended for optimal performance ***" ) );
```
probably we should disable this suggestion for now till we will be able to do so

lutovich commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/MultiPassStore.java:
```diff
@@ -164,14 +144,35 @@ public static boolean recordInCurrentPass( long id, int iPass, long recordsPerPa
         }
 
-        StoreProcessor[] processors( MultiPassStore... stores )
+        StoreProcessor[] processors( Stage stage, MultiPassStore... stores )
```
Method not used

lutovich commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/NodeLabelReader.java:
```diff
@@ -81,6 +89,37 @@ public void onWellFormedChain( long[] labelIds,
+        return NodeLabelsField.get( nodeRecord, null );
+    }
+
+    public static <RECORD extends AbstractBaseRecord, REPORT extends ConsistencyReport> Set<Long> getListOfLabels(
```
Type parameter `RECORD` not used

lutovich commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/StoreProcessor.java:
```diff
@@ -145,4 +205,61 @@ public void processSchema( RecordStore<DynamicRecord> store, DynamicRecord schem
         }
     }
+
+    public <R extends AbstractBaseRecord> void applyFilteredParallel( RecordStore<R> store,
```
This method looks very similar to `ParallelRecordScanner#scan`. Could those two be unified?

lutovich commented on community/consistency-check/src/main/java/org/neo4j/consistency/store/FilteringRecordAccess.java:
```diff
@@ -100,6 +97,20 @@ public FilteringRecordAccess( DiffRecordAccess delegate, final int iPass,
+            @Override
+            public boolean accept( PropertyRecord item )
+            {
+                return !shouldSkip( item.getId() /*for some reason we don't care about the id*/,
```
It is quite confusing

lutovich commented:
@tinwelint I'm done with review. Made some minor codestyle comments, otherwise looks good to me.

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/PropertyRecordCheck.java:
```diff
@@ -253,7 +177,15 @@ public void checkConsistency( PropertyRecord record,
             if ( !NONE.is( valueFrom( record ) ) )
             {
-                engine.comparativeCheck( records.property( valueFrom( record ) ), this );
+                PropertyRecord prop = records.cacheAccess().client().getPropertyFromCacheCache( valueFrom( record ) );
```
`getPropertyFromCacheCache` too many caches

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/RelationshipRecordCheck.java:
```diff
@@ -19,32 +19,78 @@
+        super( fields );
+    }
+
+    public static RelationshipRecordCheck relationshipRecordCheckPass1()
```
`relationshipRecordCheckPass1` and `relationshipRecordCheckPass2` should be updated with more descriptive names

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/DefaultCacheAccess.java:
```diff
@@ -0,0 +1,188 @@
+        @Override
+        public boolean withinBounds( long id )
+        {
+            return !(recordsPerCPU > 0 && (id > (threadIndex + 1) * recordsPerCPU || id < threadIndex * recordsPerCPU) );
```
bit more brackets should be helpful here

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/cache/PackedMultiFieldCache.java:
```diff
@@ -0,0 +1,105 @@
+
+    private static boolean isId( int[] slotSizes, int i )
+    {
+        return slotSizes[i] >= 35;
```
constant should be handy here

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/CheckStage.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public enum CheckStage implements Stage
+{
+    Stage1_NS_PropsLabels( false, true, "NodeStore pass - check its properties, check labels and cache them, skip relationships", 1, 63 ),
```
also here at least for 63 and 35 constants should be more descriptive

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/ConsistencyCheckTasks.java:
```diff
@@ -53,120 +60,166 @@
-            tasks.add( create( nativeStores.getRelationshipGroupStore(), multiPass.processors( RELATIONSHIP_GROUPS ) ) );
+            MandatoryProperties mandatoryProperties = new MandatoryProperties( nativeStores );
+            StoreProcessor processor =
+                    multiPass.processor( CheckStage.Stage1_NS_PropsLabels, PROPERTIES );
```
looks like if we will introduce stages hierarchy it should be much more easier to construct required tasks, wdyt?

tinwelint replied:
What do you mean?

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/FullCheck.java:
```diff
@@ -111,44 +114,38 @@ public ConsistencySummaryStatistics execute( DirectStoreAccess stores, Log log )
-
-        ProgressMonitorFactory.MultiPartBuilder progress = progressFactory.multipleParts( "Full consistency check" );
+        StoreProcessor processEverything = new StoreProcessor( decorator, reporter, Stage.SEQUENTIAL_FORWARD, cacheAccess );
+        StringBuilder consoleMsg = new StringBuilder( "Full Consistency Check" );
```
`consoleMsg` can be inlined

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/ParallelRecordScanner.java:
```diff
@@ -0,0 +1,98 @@
+      {
+          try
+          {
+              // do a round robin distribution to maintain physical locality
```
i think we need a bit more explanation here

tinwelint committed:
Brings in the legacy consistency checker as a separate component

which will live in community/consistency-check-legacy until we're
confident that the new and big changes are stable and works in all
scenarios. The legacy checker is a complete copy of the code base as it
was in 2.3 HEAD right before this commit, and as such can be deleted
without affecting the new checker when that time comes.

The legacy checker will be default for a little while longer, but can be
run if --experimental is supplied.

Up to date with commit be67faa2c0942071705b11828ca0d9753be67d36

tinwelint commented:
@MishaDemianenko @lutovich I've addressed all comments, except one of yours Misha.

davidegrohmann commented:
@tinwelint this needs a rebase
